### PR TITLE
[Behat][Promotion] Discount based on number of products from taxon

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | np
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | partially https://github.com/Sylius/Sylius/issues/4194
| License         | MIT

### CODE REVIEW

File: src/Sylius/Component/Core/spec/Factory/RuleFactorySpec.php [hunk 1]
```diff
@@ -80,4 +81,13 @@ function it_creates_total_of_items_from_taxon_rule($decoratedFactory, RuleInterf
 
         $this->createItemsFromTaxonTotal('spears', 1000)->shouldReturn($rule);
     }
+
+    function it_create_contains_taxon_tule($decoratedFactory, RuleInterface $rule)
```
Comment (NeverResponse) [inline src/Sylius/Component/Core/spec/Factory/RuleFactorySpec.php]: `*_creates_*`

File: src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    function it_returns_false_if_taxon_with_configured_code_does_not_exist(
+        OrderInterface $order,
+        TaxonRepositoryInterface $taxonRepository
+    ) {
+        $taxonRepository->findOneBy(['code' => 'bows'])->willReturn(null);
+
+        $this->isEligible($order, ['taxon' => 'bows', 'count' => 10])->shouldReturn(false);
+    }
+
+    function it_throws_exception_if_promotion_subject_is_not_order(PromotionSubjectInterface $subject)
```
Comment (NeverResponse) [inline src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php]: `*_if_the_promotion_subject_is_not_an_order(`

File: src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        $compositeBowItem->getQuantity()->willReturn(4);
+
+        $longbowItem->getProduct()->willReturn($longbow);
+        $longbow->hasTaxon($bows)->willReturn(true);
+        $longbowItem->getQuantity()->willReturn(5);
+
+        $this->isEligible($order, ['taxon' => 'bows', 'count' => 5])->shouldReturn(true);
+    }
+
+    function it_returns_false_if_order_has_not_required_number_of_products_from_configured_taxon(
```
Comment (NeverResponse) [inline src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php]: `*_the_order_doesnt_have_the_required_*`
or `an_order`, no idea if there was a discussion about treating it as one order, or any given that may come by

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+Feature: Receiving discount based on number of products from specific taxon
+    In order to pay less while buying required number of goods from promoted taxon
+    As a Customer
+    I want to receive discount for my purchase
+
+    Background:
+        Given the store operates on a single channel in "France"
+        And the store classifies its products as "T-Shirts" and "Mugs"
+        And the store has a product "PHP T-Shirt" priced at "€100.00"
+        And it belongs to "T-Shirts"
```
Comment (TheMadeleine) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: I'm wondering if it wouldn't be better to rename this step to: `And this product belongs to "T-Shirts"`
wdyt?

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        And the store has a product "Symfony T-Shirt" priced at "€150.00"
+        And it belongs to "T-Shirts"
+        And the store has a product "PHP Mug" priced at "€20.00"
+        And it belongs to "Mugs"
+        And there is a promotion "T-Shirts promotion"
+
+    @ui
+    Scenario: Receiving discount on order while buying required number of products from promoted taxon
```
Comment (TheMadeleine) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: `Scenario: Receiving a discount on order while buying required number of products from a promoted taxon`
Take care of that in the whole file. :)

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        Given the promotion gives "€50.00" off if order contains 2 products classified as "T-Shirts"
+        When I add 3 products "PHP T-Shirt" to the cart
+        Then my cart total should be "€250.00"
+        And my discount should be "-€50.00"
+
+    @ui
+    Scenario: Receiving discount on order while buying equal with required number of products from promoted taxon
```
Comment (TheMadeleine) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: `Scenario: Receiving a discount on order while buying exactly the required number of products from a promoted taxon` ?

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    @ui
+    Scenario: Receiving discount on order while buying multiple items with products from promoted taxon which fit number criteria
+        Given the promotion gives "€100.00" off if order contains 4 products classified as "T-Shirts"
+        When I add 3 products "PHP T-Shirt" to the cart
+        And I add 2 products "Symfony T-Shirt" to the cart
+        Then my cart total should be "€500.00"
+        And my discount should be "-€100.00"
+
+    @ui
+    Scenario: Receiving different discount on different promotions checking number of products from specific taxon
```
Comment (TheMadeleine) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: discounts?

File: src/Sylius/Behat/spec/Context/Setup/PromotionContextSpec.php [hunk 8]
```diff
@@ -362,4 +362,29 @@ function it_creates_fixed_discount_promotion_with_total_of_items_from_taxon_rule
 
         $this->thePromotionGivesOffIfOrderContainsProductsClassifiedAsAndPricedAt($promotion, 500, $tanks, 1000);
     }
+
+    function it_creates_fixed_discount_promotion_with_contains_taxon_rule(
```
Comment (TheMadeleine) [inline src/Sylius/Behat/spec/Context/Setup/PromotionContextSpec.php]: `it_creates_a_fixed_discount_promotion_which_contains_a_taxon_rule`

File: src/Sylius/Component/Core/spec/Factory/RuleFactorySpec.php (hunk 1 above)
Comment (TheMadeleine) [inline src/Sylius/Component/Core/spec/Factory/RuleFactorySpec.php]: even more:
`it_creates_a_contains_taxon_rule` :)

File: src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php (hunk 3 above)
Comment (TheMadeleine) [inline src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php]: `it_returns_false_if_an_order_does_not_have_the_required_number_of_products_from_a_configured_taxon`

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 9]
```diff
@@ -0,0 +1,55 @@
+@promotion
+Feature: Receiving discount based on number of products from specific taxon
```
Comment (lchrusciel) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: Receiving **a** discount based on number of products from **the** specific taxon

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    In order to pay less while buying required number of goods from promoted taxon
```
Comment (lchrusciel) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: from **the** promoted taxon? /cc @TheMadeleine

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    As a Customer
+    I want to receive discount for my purchase
```
Comment (lchrusciel) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: receive **a** discount? /cc @TheMadeleine

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature (hunk 9 above)
Comment (TheMadeleine) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: Receiving **a** discount based on **the** number of products from **a** specific taxon

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature (hunk 4 above)
Comment (Zales0123) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: I think using ``it`` when there is ``a product`` in previous step is clear enough, but I would be happy to read anyone else opinion :)

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        When I add 3 products "PHP T-Shirt" to the cart
+        And I add 2 products "Symfony T-Shirt" to the cart
+        Then my cart total should be "€500.00"
+        And my discount should be "-€100.00"
+
+    @ui
+    Scenario: Receiving different discount on different promotions checking number of products from specific taxon
+        Given the promotion gives "€50.00" off if order contains 2 products classified as "T-Shirts"
+        And there is a promotion "Mugs promotion"
+        And it gives "€10.00" off if order contains 3 products classified as "Mugs"
```
Comment (lchrusciel) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: if an order?

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature (hunk 10 above)
Comment (TheMadeleine) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: **a**

File: src/Sylius/Component/Core/Factory/RuleFactoryInterface.php [hunk 13]
```diff
@@ -48,4 +48,12 @@ public function createTaxon(array $taxons);
      * @return RuleInterface
      */
     public function createItemsFromTaxonTotal($taxon, $amount);
+
+    /**
+     * @param string $taxon
+     * @param int $count
+     *
+     * @return mixed
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Factory/RuleFactoryInterface.php]: Can't we be more precise?

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature (hunk 12 above)
Comment (Zales0123) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: :+1: but I think it should be changed in all ``And it gives`` step and it's enough work for separate PR :smile:

File: features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature (hunk 10 above)
Comment (lchrusciel) [inline features/promotion/receiving_discount_based_on_number_of_items_from_specific_taxon.feature]: Someday I will learn how to use them properly :disappointed_relieved:

File: src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+
+    function it_throws_exception_if_the_promotion_subject_is_not_an_order(PromotionSubjectInterface $subject)
+    {
+        $this
+            ->shouldThrow(new UnexpectedTypeException($subject->getWrappedObject(), OrderInterface::class))
+            ->during('isEligible', [$subject, []])
+        ;
+    }
+
+    function it_has_no_configuration_type_configured_yet()
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/spec/Promotion/Checker/ContainsTaxonRuleCheckerSpec.php]: :D

**Comment by pjedrzejewski** [comment]
Thank you Mateusz! :+1:

**Pull request merged**